Use [[nodiscard]] in places with memory leak risk

PBfordev opened the pull request:
I understand if this is rejected, considering the attribute not very useful and just adding noise to a declaration.

I believe it could be used at least in places with a memory leak risk as I did here for MSW-specific code. 

Discarding the return value from any `Get*()`, `Is*()`, (let alone `Clone()`) function almost always indicates a bug, but there are so many such functions that it is probably not realistic to add it everywhere needed.

There may be more functions with a memory leak risk but for now, I have just added the few I could think off the top of my head. If this proposal is accepted, I will look for more candidates.

**EDIT**
It seems that the attribute could be used even with many widespread pre-C++17 compilers, see https://stackoverflow.com/a/45191044/7267315.

PBfordev committed:
Use [[nodiscard]] in places with memory leak

Introduce [[nodiscard]] attribute to the code base, using wxNODISCARD
which allows the code to build with pre-C++17 compilers.

For now, use it only in several of the places with a risk of memory leak.

SpareSimian commented:
Should this be used for all singletons, like wxPersistenceManager::Get()? I use this in my wxApp subclass to ensure that the manager had the correct lifetime. (But being a singleton it probably lasts until after exit() destroys all global data, which is all I need.)

Blake-Madden commented:
You could add a `wxNODISCARD("message")` variant also for when C++20 is defined, which allows for something like this:

`[[nodiscard("Assign the return to avoid a memory leak.")]]`

Personally, I find `nodiscard` very useful; it's helped me find a couple of issues before in my own code.

PBfordev commented:
I realized that GCC and clang warning cannot be suppressed with just casting the result to `void*` as with MSVC.

This makes the warning rather annoying in the rare cases we really do not care about the return value, like in this PR when detaching the safe array we just "borrowed" with attach earlier.

I am not sure about the variant with the message. Even if this is accepted, the attribute is probably going to be used in wxWidgets sparingly, only in the cases where there is a risk of memory leak so the reason for the warning should be clear even without the message.

PBfordev committed:
Use the attribute with more compilers

Warn about the unused return value also with pre-C++17 compilers
where available.

Any supported clang or GCC version has __attribute__ ((warn_unused_result)).

With MSVC, use SAL's _Check_return_.

PBfordev commented:
I have added support for warning for more compilers.

But perhaps I should have defined the macro as  `WX_ATTRIBUTE_NODISCARD`, similarly to the already defined `WX_ATTRIBUTE_NORETURN` or `WX_ATTRIBUTE_UNUSED`...

vadz commented on include/wx/defs.h:
```diff
@@ -259,6 +259,19 @@ typedef short int WXTYPE;
     #define wxFALLTHROUGH ((void)0)
 #endif
 
+/* wxNODISCARD is used to notate that the function return value should not be ignored */
```
I'm sorry, but is "notate" the right word to use here? I'd say "indicate". I'd also use "must" instead of "should".

PBfordev replied:
I just copied the comment from `wxFALLTHROUGH`, probably wrongly.

I will change the wording to what you suggested, once we agreed on other changes.

vadz commented on include/wx/msw/ole/safearray.h:
```diff
@@ -367,7 +367,7 @@ class wxSafeArray : public wxSafeArrayBase
 
         if ( sa.HasArray() )
-            sa.Detach();
+            static_cast<void*>(sa.Detach()); // we do not own the array, just attached it
```
Could/should we use `wxUnusedVar()` instead of the cast? It would seem to be more clear, although the most clear would probably be to rename that function (while keeping the current name for compatibility, of course...) to `wxIgnoreUnusedWarning()` as then the code using it would really be self-documenting.

PBfordev replied:
As I wrote elsewhere, I realized that the cast cannot be used since the warning is produced with clang and GCC.

I think for now I am going to use `wxUnusedVar()` and the renaming of the function should be left to a different PR.

vadz replied:
I did see that you write it, but then I've also seen that the CI builds passed and so thought you've fixed it by using static cast instead of a plain cast or something like this. If not, I don't understand why did they pass, they're supposed to use `-Werror` and fail if there are any warnings... Do you understand what's going on here or should I look at this?

PBfordev replied:
(Sorry, I missed this comment before).

I did not not know that (some?) CI builds do not allow any warnings. I get a few between the library and samples when building with GCC locally (at least when using CMake).

I am ashamed to admit I cannot reproduce the GCC warning when reverting from `wxUnusedVar` to `static_cast<void*>` anymore, I can't believe that extending the macro definition would change that (tried with both 11 and 17 C++ standards). But it seems that generally the warnings are always not the same (depends on C++ standard, e.g., the string template one from the XML library ), but perhaps I just miss them...

vadz reviewed (commented):
Thanks, this is definitely going to be useful!

I can merge it already (with just a tweak to the comment) if you prefer, or you could replace the casts if you agree that it would be better.

FWIW I also like the idea of having `wxNODISCARD_MSG()`, but it's not critical and could always be added later, of course.

PBfordev commented:
> I can merge it already (with just a tweak to the comment) if you prefer, or you could replace the casts if you agree that it 

So the macro should be left as is now and we should not name it use `WX_ATTRIBUTE_NODISCARD` (analogical to already defined `WX_ATTRIBUTE_NORETURN` or `WX_ATTRIBUTE_UNUSED` in the same file)?

vadz commented:
I prefer the current name, I think the `WX_ATTRIBUTE_XXX` ones predate the standard attributes addition and `wxNODISCARD` is consistent with `wxFALLTHROUGH` if not with them.

PBfordev committed:
!fixup

1. Fix the comment describing the macro.
2. Use wxUnusedVar() instead of a cast to prevent the nodiscard warning

PBfordev commented:
BTW, there is a word missing in the initial commit title, it should read Use [[nodiscard]] in places with memory leak **risk**

But perhaps the commit title should be changed entirely, e.g., replace "in places" with "in functions".
